### DISCUSSION TITLE
Are there ways to speed up preview for large qmd/html files?

### DISCUSSION THREAD

**Discussion by jmhimara** [quarto-dev/quarto-cli#5532] [Q&A]
This might be a silly question, but I'm working with some large qmd files (~6000 loc), and the preview functionality on vscode is quite slow (about 30 seconds per change). It's not a huge deal, but bit inconvenient.  There's no python or R code, just html and a little bit of javascript. 

Is there any way to speed up the preview update?  Or perhaps split the qmd file into multiples but still render it as a single html?

Thanks!

**Comment by cscheid** [comment]
Can you share your file? We're working on render performance for 1.4 and it'll help if we know what's triggering the issue for you.

**Reply by jmhimara** [reply]
Yeah, sure. It's part of a book, but here's the chapter that's taking the longest to reload (the rep contains the rest of the book, if that's relevant): 

https://github.com/jmhimara/bfrpg/blob/main/monstersAll.qmd

**Reply by cscheid** [reply]
Thanks. I see it. The issue is that this is actually really just a big file, and our AST processing chain is not very efficient at the moment.

If you don't mind, I'll keep this file around as a benchmark and use it to measure the work we are doing. I don't have anything immediate to help you with at the moment, but as we work on this I'll come back to the discussion.

**Comment by Hakarl22** [comment]
This is more of a general question regarding writing books or large documents in Quarto.  What are the size limits?  I have been working on a book for a few years using quarto now and keep running into issues with speed and hang-ups.  I would love to see some general guidelines to follow.  For example if you are running large datasets then should you run them from within quarto or in a separate documents?  I am an experienced LaTeX user and hoped to see a similar performance from Quarto.  I know that LaTeX is only text but still it would be nice to have some guidelines.  Clearly if you cannot run code - even on a large dataset inside quarto - then it is no longer truely reproducible.  Hope to get some advice.

Thanks,
H

**Reply by cderv** [reply]
> What are the size limits? I have been working on a book for a few years using Quarto now and keep running into issues with speed and hang-ups.

AFAIK the situation will depend on the environment where rendering happens, the size of the data and the content of the document. Regarding execution, Quarto won't improve what takes long to run outside of Quarto. So if computation limits the rendering, work needs to be done on that front using tools available for each language. It could be caching long computation, or using tools to distribute computation. Quarto won't do it on its own, but it can be coupled with such workflow (in the R ecosystem, there is an example of **targets** package with its [quarto integration](books.ropensci.org/targets/literate-programming.html#quarto-targets))

If the limitation happens after computation, know that the next release of 1.7 already has several speed improvements and that we are constrained by the toolchain: Pandoc is used to produce the text file, and LaTeX is used to render to PDF. If the limitation is at LaTeX rendering, Quarton won't be able to do much. Anything after computation and output production, we'll keep improving. 

Anyway, those are general comments. Without knowing more about your use case, it is hard to recommend and pinpoint the precise unlock to your current limitations.
